ui/src/serverIO.js

```diff
@@ -376,27 +376,24 @@ class ServerIO {
       this.dispatch(showResumeQueueDialog(true));
     });
 
-    this.hwrSocket.on('observersChanged', (data) => {
-      const state = store.getState();
+    this.hwrSocket.on('userChanged', async (message) => {
+      const prevState = store.getState();
+      const { inControl: wasInControl } = prevState.login.user;
 
-      if (
-        data.observers.length > 0 &&
-        data.operator.username === state.login.user.username &&
-        !state.login.user.inControl
-      ) {
-        this.dispatch(showWaitDialog('You were given control', data.message));
-      } else if (
-        data.observers.length > 0 &&
-        state.login.user.inControl &&
-        data.observers
-          .map((el) => el.username)
-          .includes(state.login.user.username)
-      ) {
+      await this.dispatch(getLoginInfo());
+
+      const newState = store.getState();
+      const { inControl } = newState.login.user;
+
+      if (!wasInControl && inControl) {
+        this.dispatch(showWaitDialog('You were given control', message));
+      } else if (wasInControl && !inControl) {
         this.dispatch(showWaitDialog('You lost control'));
       }
```

Comment:
This is the new logic to show the "You were given control" and "You lost control" dialog. It looks only for a change in the `login.user.inControl` state — the server no longer needs to send a payload with the `userChanged` message (apart from an optional `message` to display in the dialog).

I've done my best to test this and can't seem to find a case where the new logic behaves differently from the old logic...

The key was to `await` the refetch of the `login` state and to compare the values before and after. I will use the same pattern to show/hide the "asking for control" dialog and the like in future PRs.

Reply:
Very nice, thanks alot. We will have to make some thorough testing of this during startup but its well worth it :)